Add programming section to docs (#782)
pasqal-io/Pulser: docs/source/programming.md

> # Programming a neutral-atom Quantum Computer

Author: But do people know what QPU stands for ?

> # Programming a neutral-atom Quantum Computer

Reviewer: I'd imagine so but maybe I'm wrong

## Introduction

1. Atoms encode the state

Neutral atoms store the quantum information in thier energy levels. If only 2 energy levels of an atom are being used, the atom is a qubit. If these eigenstates are $\left|a\right>$ and $\left|b\right>$, then the state of the atom is described by $\left|\psi\right> = \alpha \left|a\right> + \beta \left|b\right>$, with $|\alpha|^2 + |\beta|^2 = 1$.

> When multiple atoms are used, the state describing this multi-atom system is the tensor network of the states. Say these atoms are labeled $\[1, 2, ..., N\]$, then the state of the system is $\left|\Psi\right> = \left|\psi_1\right> \otimes \left|\psi_2\right> \otimes ... \otimes \left|\psi_N\right>$.

Author: I agree, I based myself on this to propose a new one: https://dojo.qulacs.org/en/latest/notebooks/1.3_multiqubit_representation_and_operations.html

> 2. Hamiltonian evolves the state

Author: This is not a nit haha I wanted to change this ;)

In quantum physics, the state of a system of atoms evolve along time following the Schrödinger equation: $i\frac{d\left|\Psi\right>(t)}{dt} = \frac{H(t)}{\hbar} \left|\Psi\right>(t)$. $H(t)$ is the Hamiltonian describing the evolution of the system. **Pulser provides you the tools to program this Hamiltonian**. 

> 2.1. The Hamiltonian

Reviewer: But then you have 2.1 twice

The Hamiltonian describing the evolution of the system can be written as

$$
H(t) = \sum_i \left (H^D_i(t) + \sum_{j<i}H^\text{int}_{ij} \right),
$$

where $H^D_i$ is the driving Hamiltonian for atom $i$ and
$H^\text{int}_{ij}$ is the interaction Hamiltonian between atoms $i$
and $j$.

2.1. Driving Hamiltonian

The driving Hamiltonian describes the effect of a pulse of Rabi frequency $\Omega(t)$, detuning $\delta(t)$ and phase $\phi(t)$ on an individual atom, exciting the transition
between two of its energies levels, $|a\rangle$ and $|b\rangle$.

:::{figure} files/two_level_ab.png
:align: center
:alt: The energy levels for the driving Hamiltonian.
:width: 200

$$
H^D(t) / \hbar = \frac{\Omega(t)}{2} e^{-i\phi(t)} |a\rangle\langle b| + \frac{\Omega(t)}{2} e^{i\phi(t)} |b\rangle\langle a| - \delta(t) |b\rangle\langle b|

> $$

Author: ![image](https://github.com/user-attachments/assets/5f6d27ce-712c-4456-92f9-1d9aecf81c8c)

![image](https://github.com/user-attachments/assets/786df260-e131-4f0e-9db7-5336fd11e5f2)

![image](https://github.com/user-attachments/assets/498598c3-6e6c-4130-ba94-b376586bc2d0)

> $$

Author: Note: perhaps we should delete the figure with the transition.

2.2. Interaction Hamiltonian

> The interaction Hamiltonian depends on the distance between the atoms $i$ and $j$, $R_{ij}$, and the energy levels in which the information is encoded in these atoms, that define the interaction between the atoms $\hat{E}_{ij}$

Author: Sure, I didn't check and named it E for entanglement 🙄

$$
H^\text{int}_{ij} = \hat{E}_{ij}(R_{ij})
$$

The interaction operator $\hat{E}_{ij}$ is composed of an entangling operator and an interaction strength:
- If the rydberg state $\left|r\right>$ is involved in the computation, then $\hat{E}_{ij}(R_{ij}) = \frac{C_6}{R_{ij}^6} |r\rangle\langle r|_i |r\rangle\langle r|_j$, where the interaction strength is $\frac{C_6}{R_{ij}^6}$, with $C_6$ a coefficient that can be programmed using Pulser. The entangling operator between atom $i$ and $j$ is $\hat{n}_i\hat{n}_j = |r\rangle\langle r|_i |r\rangle\langle r|_j$.
- If the information is stored in the $\left|0\right>$ and $\left|1\right>$ states of the `XY` basis, $\hat{E}_{ij}(R_{ij}) =\frac{C_3}{R_{ij}^3} (|1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j)$, where the interaction strength is $\frac{C_3}{R_{ij}^3}$, with $C_3$ a coefficient that can be programmed using Pulser. The entangling operator between atom $i$ and $j$ is $\hat{\sigma}_i^{+}\hat{\sigma}_j^{-} + \hat{\sigma}_i^{-}\hat{\sigma}_j^{+} = |1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j$.

> 

Author: What about:
```suggestion
:::{important}
With Pulser, you program the interaction Hamiltonian by setting the distance between atoms, $R_{ij}$, and the choice of eigenstates used in the computation and the Rydberg level(s) targeted by the device. Most commonly, the ground and Rydberg states ($\left|g\right>$ and $\left|r\right>$) are used, such that
$$
H^\text{int}_{ij} = \frac{C_6}{R_{ij}^6} \hat{n}_i\hat{n}_j
$$
When providing the distance between atoms, Pulser ensures that you respect the constraints of your chosen device.
:::
```

> 

Author: Just to be closer to the :::important section for the driving hamiltonian

> 

Reviewer: I'm ok with that, but the first sentence needs to be rephrased, I find it quite confusing. How about this:
```
With Pulser, you program the interaction Hamiltonian by setting the distance between atoms, $R_{ij}$. Additionally, the choice of eigenstates used in the computation and the Rydberg level(s) targeted by the device fully determine the interaction strength.
```

> 

Reviewer 2: With Pulser, you program the interaction Hamiltonian by setting the _relative_ distance between atoms, $R_{ij}$. 

The choice of the eigenstates used in the computation and therefore the Rydberg level(s) targeted is done by the user by selecting a specific device (link to section?) after inspecting its specifications. 

Most commonly, the ground and Rydberg states ($\left|g\right>$ and $\left|r\right>$) are used, such that

> 

Author: I have added in the XY Hamiltonian that the |0> and |1> states are Rydberg states, because otherwise people would have gotten lost. @ferrulli1pasqal, I won't say _relative_ distance because it gives the impression that it's between 0 and 1 (relative to what ?) The chose of the eigenstates in the computation is not done by selecting the device, but by declaring the channels (and adding pulses to them, if you don't add any pulse it does not count).

2.3. Evolution of the system's state

For a system of atoms in the initial state $\left|\Psi_0\right>$, the final state of the system after a time $\Delta t$ is:

> $$ \left|\Psi_f\right> = \exp\left(-\frac{i}{\hbar}\int_0^{\Delta t} H(t) dt\right)$$

Author: Personnally, that's how I "program" the final state of my system, knowing the initial state. I wanted to add it at the beginning of the section 2, but then I thought that would be too much with the Schrödinger equation, I prefered having it as the last section to remember the user why we have detailed the hamiltonian terms just before

## Recipe

> As underlined above, Pulser lets the user program an Hamiltonian to modify the state of a system of atoms, that encode the states. What do you define at each step of a quantum program in Pulser ? Let's go over these steps.

Author: I am only changing the last sentence to "Here is a step-by-step guide to create your own Pulser `Sequence`."

1. Pick a Device

```{mermaid}

flowchart TB
  A[Picking a Device] --> B{Do I want to run on a QPU}
  B -->|Yes| C[Pick the Device from the [QPU backend](./tutorials/backends.nblink)]
  B -->|No| D[Do I want to mimic all the QPU's constraints ?]
  D -->|Yes| E[Use a `Device` (like `pulser.AnalogDevice`)]
  D --> |No| F[Use a [`VirtualDevice`](./tutorials/virtual_devices.nblink) (like `pulser.MockDevice`)]

```

> The `Device` you select will dictate some parameters and constraint others. For instance, the value of the $C_6$ and $C_3$ coefficients of the Interaction Hamiltonian are defined by the device. For a complete description of the constraints introduced by the device, [check its description](./apidoc/core.rst).

Author: Ah yes, I redirected to the api doc only because the description page is not ready yet, I will do this afterwards

2. Create the Register

> The Register defines the position of the atoms. This determines:

Reviewer: ```suggestion
The `Register` defines the position of the atoms. This determines:
```

- the number of atoms to use in the quantum computation, i.e, the size of the system (let's note it $N$).
- the distance between the atoms, the $R_{ij} (1\le i, j\le N)$ parameters in the interaction Hamiltonian.

> At this stage, the interaction strength of the Hamiltonian is fully determined.

Author: That's true, I don't know what I wanted to say ^^'

> At this stage, the interaction strength of the Hamiltonian is fully determined.

Author: I think I just mistook where I wanted to put this, I moved it to the next section.

> 3. Pick the Channels

Author: Aren't local channels for the advanced part ? Otherwise I am fine with adding them

This defines the energy levels that will be used in the computation. A Channel targets the transition between two energy levels. The Channel must be picked from the `Device.channels`: select your device for it to have the channels you need.

Picking the channel will initialize the state of the system, and fully determine the Interaction Hamiltonian:
- If the selected Channel is the `Rydberg` or the `Raman` channel, the system is initialized in $\left|gg...g\right>$ and the interaction Hamiltonian is $H^\text{int}_{ij} =\frac{C_6}{R_{ij}^6}|r\rangle\langle r|_i |r\rangle\langle r|_j$.

> - If the selected Channel is the `Microwave` channel, the system is initialized in $\left|11...1\right>$ and the interaction Hamiltonian is $H^\text{int}_{ij} =\frac{C_3}{R_{ij}^3}|1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j$.

Author: Daaamn

# Programming a neutral-atom Quantum Computer

## Introduction

> 

Reviewer 2: how about having in every page the list of "what will you learn in this page".

This can help the user a lot and probably be very helpful for indexing pages for search engines, like typing in google "rydberg" would bring us up in the search. Wdyt?

Author: Do you really think it will help for page indexation in Google ? If you have documentation regarding this I am all hear 
I have not found better than this https://blog.readthedocs.com/seo-for-technical-docs/
But they don't mention this. However, I guess it can help the user so let's try to write something indeed.

Reviewer 2: So my starting point was to facilitate the users and I think also the structure of the documentation. In principle by just looking at this first box for each page we should be able to see what we cover and what we don't.
The indexing was a bit speculative: I don't know any detail about how this could work but at the same point I am pretty confident that this should help in this direction.
Probably the best people to ask as frontend people. I iwll drop a message askingif anyone has good advice in this direction too

### 1. Atoms encode the state

Neutral atoms store the quantum information in their energy levels (also known as [_eigenstates_](./conventions.md)). When only two eigenstates are used to encode information, each atom is a qubit. If these eigenstates are $\left|a\right>$ and $\left|b\right>$, then the state of an atom is described by $\left|\psi\right> = \alpha \left|a\right> + \beta \left|b\right>$, with $|\alpha|^2 + |\beta|^2 = 1$.

> When multiple atoms are used, the state of the system is described by a linear combination of the _eigenstates_ of the multi-atom system, whose set is obtained by making the cross product of the set of eigenstate of each atom. If each atom is described by $d$ eigenstates labelled ${\left|a_1\right>, \left|a_2\right>...\left|a_d\right>}$ (each atom is a _qudit_), and if there are $N$ atoms in the system, then the state of the whole system is provided by $\left|\Psi\right> = \sum_{i_1, ..., i_N \in [1, ..., d]} c_{i_1, ..., i_N} \left|a_{i_1}...a_{i_N}\right>$, with $\sum_{i_1, ..., i_N \in [1, ..., d]} |c_{i_1, ..., i_N}|^2 = 1$.

Reviewer 2: avoid inline equation for long equations to help the reader

### 2. Hamiltonian evolves the state

In quantum physics, the state of a quantum system evolves along time following the Schrödinger equation: $i\frac{d\left|\Psi\right>(t)}{dt} = \frac{H(t)}{\hbar} \left|\Psi\right>(t)$, where $H(t)$ is the Hamiltonian describing the evolution of the system. 

**Pulser provides you the tools to program this Hamiltonian**. 

#### 2.1. The Hamiltonian

The Hamiltonian describing the evolution of the system can be written as

$$
H(t) = \sum_i \left (H^D_i(t) + \sum_{j<i}H^\text{int}_{ij} \right),
$$

where $H^D_i$ is the driving Hamiltonian for atom $i$ and
$H^\text{int}_{ij}$ is the interaction Hamiltonian between atoms $i$
and $j$.

#### 2.2. Driving Hamiltonian

> The driving Hamiltonian describes the effect of a pulse of Rabi frequency $\Omega(t)$, detuning $\delta(t)$ and phase $\phi(t)$ on an individual atom, driving the transition

Reviewer 2: I would split this sentence in two, like:
Two drive each atom excitation between the two eigenstates, neautral atom QPUs use laser beams. In partcular, ...[Rabi description you provided]

> The driving Hamiltonian describes the effect of a pulse of Rabi frequency $\Omega(t)$, detuning $\delta(t)$ and phase $\phi(t)$ on an individual atom, driving the transition

Reviewer 2: Probably, somewhere in conventions or notation we can have explicited the relation between detuning and phase as you explained to me. This can be helpful to explain later in other sections the choice made in the code to provide the detuning only.

between two of its energies levels, $|a\rangle$ and $|b\rangle$.

:::{figure} files/two_level_ab.png
:align: center
:alt: The energy levels for the driving Hamiltonian.
:width: 200

The coherent excitation is driven between a lower energy level, $|a\rangle$, and a higher energy level,
$|b\rangle$, with Rabi frequency $\Omega(t)$ and detuning $\delta(t)$.
:::

$$
H^D(t) / \hbar = \frac{\Omega(t)}{2} e^{-i\phi(t)} |a\rangle\langle b| + \frac{\Omega(t)}{2} e^{i\phi(t)} |b\rangle\langle a| - \delta(t) |b\rangle\langle b|
$$

:::{important}
With Pulser, you program the driving Hamiltonian by setting $\Omega(t)$, $\delta(t)$ and $\phi(t)$, all the while Pulser ensures that you respect the constraints of your chosen device.
:::

#### 2.3. Interaction Hamiltonian

> The interaction Hamiltonian depends on the distance between the atoms $i$ and $j$, $R_{ij}$, and the energy levels in which the information is encoded in these atoms, that define the interaction between the atoms $\hat{U}_{ij}$

Reviewer 2: Probably an introduction sentence where we say that:
- interaction is really the interaction forces between two atoms 
- it does not depend on time and so it is always on, contrarily to the drive part
Then you can continue with what you have written

Reviewer 2: regarding the n operator, maybe it would be worth to have it properly defined somewhere. it can he here or somewhere else.
link n_i = product tensor of I and Pauli matrix.
Something along the line as sigma_i^x= I ⊗ I ⊗ ... ⊗ I ⊗ sigma^x ⊗ I ⊗ ... ⊗ I

Author: From your first comment: I am picking your second point, because it's highly relevant ! Regarding the first one, I think it's related to the physics and I will leave this to the convention page/community page.

$$
H^\text{int}_{ij} = \hat{U}_{ij}(R_{ij})
$$

> The interaction operator $\hat{U}_{ij}$ is composed of an entangling operator and an interaction strength:

Reviewer 2: is defined by an entangling operator?

Reviewer 2: sorry, I just meant to say to use "defined by" instead of "composed"

Author: Then I don't think it's "defined by", but "composed of" haha It's defined by some factors, but it's composed of a matrix and a scalar

> - If the Rydberg state $\left|r\right>$ is involved in the computation, then $\hat{U}_{ij}(R_{ij}) = \frac{C_6}{R_{ij}^6} \hat{n}_i \hat{n}_j$, where the interaction strength is $\frac{C_6}{R_{ij}^6}$, with $C_6$ a coefficient that depends on the principal quantum number of the Rydberg state. The entangling operator between atom $i$ and $j$ is $\hat{n}_i\hat{n}_j = |r\rangle\langle r|_i |r\rangle\langle r|_j$. Together with the driving Hamiltonian, this interaction encodes the _Ising Hamiltonian_ and is the **most common choice in neutral-atom devices.**

Reviewer 2: if the chosen basis is the rydberg one (possible add a link to where the basis is explained), then...

ALso for the sake of readibility I would avoid using the bullet point here and use it instead to point the different information contained in the whole sentence.

> - If the information is stored in the $\left|0\right>$ and $\left|1\right>$ states of the `XY` basis, $\hat{U}_{ij}(R_{ij}) =\frac{C_3}{R_{ij}^3} (|1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j)$, where the interaction strength is $\frac{C_3}{R_{ij}^3}$, with $C_3$ a coefficient that depends on the energy levels used to encode $\left|0\right>$ and $\left|1\right>$. The entangling operator between atom $i$ and $j$ is $\hat{\sigma}_i^{+}\hat{\sigma}_j^{-} + \hat{\sigma}_i^{-}\hat{\sigma}_j^{+} = |1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j$. This interaction hamiltonian is associated with the _XY Hamiltonian_ and is a less common mode of operation, usually accessible only in select neutral-atom devices.

Reviewer 2: seee comment above for rydberg basis

> - If the information is stored in the $\left|0\right>$ and $\left|1\right>$ states of the `XY` basis, $\hat{U}_{ij}(R_{ij}) =\frac{C_3}{R_{ij}^3} (|1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j)$, where the interaction strength is $\frac{C_3}{R_{ij}^3}$, with $C_3$ a coefficient that depends on the energy levels used to encode $\left|0\right>$ and $\left|1\right>$. The entangling operator between atom $i$ and $j$ is $\hat{\sigma}_i^{+}\hat{\sigma}_j^{-} + \hat{\sigma}_i^{-}\hat{\sigma}_j^{+} = |1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j$. This interaction hamiltonian is associated with the _XY Hamiltonian_ and is a less common mode of operation, usually accessible only in select neutral-atom devices.

Author: Create ##### section with Ising Hamiltonian and XY Hamiltonian

:::{important}
With Pulser, you program the interaction Hamiltonian by setting the distance between atoms, $R_{ij}$, and the choice of eigenstates used in the computation and the Rydberg level(s) targeted by the device. Most commonly, the ground and Rydberg states ($\left|g\right>$ and $\left|r\right>$) are used, such that
$$
H^\text{int}_{ij} = \frac{C_6}{R_{ij}^6} \hat{n}_i\hat{n}_j
$$
When providing the distance between atoms, Pulser ensures that you respect the constraints of your chosen device.
:::

#### 2.4. Evolution of the system's state

For a system of atoms in the initial state $\left|\Psi_0\right>$, the final state of the system after a time $\Delta t$ is:
$$ \left|\Psi_f\right> = \exp\left(-\frac{i}{\hbar}\int_0^{\Delta t} H(t) dt\right)$$

## Writing a Pulser program

> As outlined above, Pulser lets you program an Hamiltonian so that you can manipulate the quantum state of a system of atoms, that encode the states. The series of necessary instructions is encapsulated in the so-called Pulser `Sequence`. Here is a step-by-step guide to create your own Pulser `Sequence`.

Reviewer 2: an Hamiltonian -> the Hamiltonian presented above.

so that you -> so that the user 

I would maybe delete "that encode the states" as redundant and explained above

### 1. Pick a Device

```{mermaid}

flowchart TB
  A[Picking a Device] --> B{Do I want to run on a QPU}
  B -->|Yes| C[Pick the Device from the QPU backend]
  B -->|No| D[Do I want to be constrained by all the QPU specs ?]
  D -->|Yes| E[Use a `Device` (like `pulser.AnalogDevice`)]
  D --> |No| F[Use a `VirtualDevice` (like `pulser.MockDevice`)]

```

> The `Device` you select will dictate some parameters and constrain others. For instance, the value of the $C_6$ and $C_3$ coefficients of the interaction Hamiltonian are defined by the device. For a complete view of the constraints introduced by the device, [check its description](./apidoc/core.rst).

Reviewer 2: Before the "For a complete view..." I would add a short but compelling sentence saying
"We strongly advice the user to have a look at the different device specifications to have a complete view of what can be achieved with Pulser. "

Author: What about adding this in the conclusion ? I don't want to scare anyone/make them visit the whole documentation at this stage of the recipe. Moreover, I currently don't have the page about the device done and I won't advice them to have a look at the API documentation at the first stage 😅

Reviewer 2: good idea to have at the end.

imho you should write in terms of what we plan to have and not what we only currently have. Otherwise you'll have always to go back to what you have written and chances are we forget.
It is a suggestion, feel free to adopt the strategy you prefer of course. It is somehow related to the discussion we were having yesterday about having a clean state new doc or to implement changes on the fly in the currently existing version. I would opt for the first for the main reasons of making more easy to read and possibly share for intermediate feedbacks, but it's your call.

### 2. Create the Register

The `Register` defines the position of the atoms. This determines:

- the number of atoms to use in the quantum computation, i.e, the size of the system (let's note it $N$).
- the distance between the atoms, the $R_{ij} (1\le i, j\le N)$ parameters in the interaction Hamiltonian.

> 

Reviewer 2: Let's briefly discuss about the opportyunity of presenting the Layout concept as well here

### 3. Pick the Channels

> A `Channel` targets the transition between two energy levels. Therefore, picking channels defines the energy levels that will be used in the computation. The channels must be picked from the `Device.channels`, so your device selection should take into account the channels it supports.

Reviewer 2: a brief reference to the driving hamiltonian would make a good transition between pulser object and the hamiltonian presented above

Author: The fact is that the channel does not only refer to the drive hamiltonian but to both the interaction and the drive hamiltonian. This is what I explain in the next paragraph. And the association channel<-> drive hamiltonian takes me a whole paragraph to develop.

Reviewer 2: Indeed...sorry I tried to look smart and comment in a rush

Picking the channel will initialize the state of the system, and fully determine the interaction Hamiltonian:
- If the selected Channel is the `Rydberg` or the `Raman` channel, the system is initialized in $\left|gg...g\right>$ and the interaction Hamiltonian is $H^\text{int}_{ij} =\frac{C_6}{R_{ij}^6}|r\rangle\langle r|_i |r\rangle\langle r|_j$.
- If the selected Channel is the `Microwave` channel, the system is initialized in $\left|00...0\right>$ and the interaction Hamiltonian is $H^\text{int}_{ij} =\frac{C_3}{R_{ij}^3}|1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j$.

> Note: It is possible to pick a `Rydberg` and a `Raman` channel, the information will then be encoded in 3 energy levels $\left|r\right>$, $\left|g\right>$ and $\left|h\right>$: the atoms are no-longer qubits but qutrits. However, it is not possible to pick the `Microwave` channel and another channel.

Reviewer 2: I would probably just mention that more complex choices in terms of channels combination can be done by the user and leave this discussion into the channel section.

Author: Or just say that `Microwave` cannot be declared along with another channel ?

Reviewer 2: I think better to just not enter into details and reserve this to the channel secction, cause I think eitherway what is said here we will have to say it again there

Author: Deleting this Note indeed.

When multiple atoms are used, the state of the system is described by a linear combination of the _eigenstates_ of the multi-atom system, whose set is obtained by making the cross product of the set of eigenstate of each atom. If each atom is described by $d$ eigenstates labelled ${\left|a_1\right>, \left|a_2\right>...\left|a_d\right>}$ (each atom is a _qudit_), and if there are $N$ atoms in the system, then the state of the whole system is provided by 

$$
\begin{align}
\left|\Psi\right> &= \sum_{i_1, ..., i_N \in [1, ..., d]} c_{i_1, ..., i_N} \left|a_{i_1}...a_{i_N}\right> \\
&= c_{1, 1, ..., 1}\left|a_{1}a_{1}...a_{1}\right> + ... + c_{1, 1, ..., d}\left|a_{1}a_{1}...a_{d}\right> + ... + c_{d, d, ..., d}\left|a_{d}a_{d}...a_{d}\right>
\end{align}
$$

where $\sum_{i_1, ..., i_N \in [1, ..., d]} |c_{i_1, ..., i_N}|^2 = 1$. If $d=2$, then this becomes

$$
\left|\Psi\right> = c_{1, 1, ..., 1}\left|a_{1}a_{1}...a_{1}\right> + c_{1, 1, ..., 2}\left|a_{1}a_{1}...a_{2}\right> + ... + c_{2, 2, ..., 2}\left|a_{2}a_{2}...a_{2}\right>
$$

If $d=2$ and $N=1$, you have the state of a qubit as above $\left|\Psi\right> = c_{1}\left|a_{1}\right> + c_{2}\left|a_{2}\right>$.

> 

Author: I am putting a figure, you will tell me if it's speaking. Perhaps we are just missing a figure in the conventions page showing all the energy levels @HGSilveri

$$
H^D(t) / \hbar = \frac{\Omega(t)}{2} e^{-j\phi(t)} |a\rangle\langle b| + \frac{\Omega(t)}{2} e^{j\phi(t)} |b\rangle\langle a| - \delta(t) |b\rangle\langle b|
$$

In the Bloch sphere representation, this Hamiltonian describes a rotation around the axis **$\Omega$**$(t) = (\Omega(t)\cos(\phi), -\Omega(t)\sin(\phi), -\delta(t))^T$, with angular velocity $\Omega_{eff}(t) = |$**$\Omega$**$(t)| = \sqrt{\Omega^2(t) + \delta^2(t)}$.

:::{figure} files/bloch_rotation_a_b.png
:align: center
:alt: Representation of the drive Hamiltonian's dynamics as a rotation in the Bloch sphere.
:width: 300

Representation of the drive Hamiltonian's dynamics as a rotation in the Bloch sphere. The coherent excitation is driven between a lower energy level, $|a\rangle$, and a higher energy level,
$|b\rangle$, with Rabi frequency $\Omega(t)$ and detuning $\delta(t)$.
:::

:::{important}
With Pulser, you program the driving Hamiltonian by setting $\Omega(t)$, $\delta(t)$ and $\phi(t)$, all the while Pulser ensures that you respect the constraints of your chosen device.
:::

> 

Reviewer 2: Is there a reason for not introducing also the formula written with the Pauli matrix representation here? (similar to what we  do in https://pulser--782.org.readthedocs.build/en/782/review.html#rydberg-states ?)

Author: Yes, because we are independent of the expression of |a> and |b> in the computational basis, the only hypothesis here is that energy level b is higher than a. It's explained in the [conventions page](https://pulser.readthedocs.io/en/stable/conventions.html#pauli-matrix-form). Overall, the conventions page is more recent, more complete and clearer than the page you are pointing at.

:::{important}
With Pulser, you program the interaction Hamiltonian by setting the distance between atoms, $R_{ij}$. Additionally, the choice of eigenstates used in the computation and the Rydberg level(s) targeted by the device fully determine the interaction strength. Most commonly, the ground and Rydberg states ($\left|g\right>$ and $\left|r\right>$) are used, such that   
$H^\text{int}_{ij} = \frac{C_6}{R_{ij}^6} \hat{n}_i\hat{n}_j$   
When providing the distance between atoms, Pulser ensures that you respect the constraints of your chosen device.
:::

## Writing a Pulser program

As outlined above, Pulser lets you program an Hamiltonian ([the Hamiltonian $H$](programming.md#2-hamiltonian-evolves-the-state)) so that you can manipulate the quantum state of a system of atoms. The series of necessary instructions is encapsulated in the so-called Pulser `Sequence`. Here is a step-by-step guide to create your own Pulser `Sequence`.

### 1. Pick a Device

:::{figure} files/decision_diagram_device.png
:align: center
:alt: Decision Diagram to select a Device for the computation
:width: 600
:::

> The `Device` you select will dictate some parameters and constrain others. For instance, the value of the $C_6$ and $C_3$ coefficients of the [interaction Hamiltonian](programming.md#22-interaction-hamiltonian) are defined by the device. For a complete view of the constraints introduced by the device, [check its description](./apidoc/core.rst).

Reviewer 2: Instead of referring to the C constant which in the end is not to crucial, I would say here that different devices would support different channels and therefore they would allow different hamiltonian implementation.

Also, the link brings the reader to the general API reference. Probably we should instead point the user to the section of the user manual where we talk about the implementation of analog and digital analog devices in pulser.devices?

Author: About your second point, I am waiting to write these pages to refer to them, but it's the idea yes :)

# Programming a neutral-atom QPU

> Pulser enables to program Quantum Processing Units (QPUs) based on neutral atoms (also named _cold-atom Quantum Processing Units_). In this page, you will learn:
> 
> - **What are the mathematical objects your are programming with Pulser ?** In quantum computing, the evolution of a quantum state is defined by an Hamiltonian. What is a quantum state in Pulser ? What are the Hamiltonians that can be defined in Pulser ?
> - **How do you program these mathematical objects ?** How do you define the quantum state in Pulser ? How do you define the Hamiltonian ? We will propose you a receipe to design your quantum program using Pulser. 

Reviewer: nit: Some suggestions, take whichever ones you like
```suggestion
Pulser establishes a framework for programming Quantum Processing Units (QPUs) based on neutral atoms (also named _cold-atom Quantum Processing Units_). In this page, you will learn:

- **What mathematical objects encode a program in a neutral-atom QPU:** In quantum computing, the evolution of a quantum state is defined by an Hamiltonian. We define both these terms, as well as the level of control Pulser provides over each one.
- **How to program these mathematical objects with Pulser:** We give you a step-by-step guide on how to create a quantum program using Pulser.
```

Reviewer: Also, the **bold** part of each bullet point could also link to the different sections in the page

## Introduction

### 1. Atoms encode the state

<details>

>   <summary>Click to expand</summary>

Reviewer: Then you can rename this as 
```suggestion
  <summary>Definition of a quantum state with multiple atoms</summary>
```

> Neutral atoms store the quantum information in their energy levels (also known as [_eigenstates_](./conventions.md)). When only two eigenstates are used to encode information, each atom is a qubit. If these eigenstates are $\left|a\right>$ and $\left|b\right>$, then the state of an atom is described by $\left|\psi\right> = \alpha \left|a\right> + \beta \left|b\right>$, with $|\alpha|^2 + |\beta|^2 = 1$.

Reviewer: I think this paragraph would be better off outside of the expandable section

<details>

  <summary>Click to expand</summary>

The driving Hamiltonian describes the effect of a pulse on two of energies levels of an individual atom, $|a\rangle$ and $|b\rangle$. A pulse is determined by its Rabi frequency $\Omega(t)$, its detuning $\delta(t)$ and its phase $\phi(t)$.

$$
H^D(t) / \hbar = \frac{\Omega(t)}{2} e^{-j\phi(t)} |a\rangle\langle b| + \frac{\Omega(t)}{2} e^{j\phi(t)} |b\rangle\langle a| - \delta(t) |b\rangle\langle b|
$$

In the Bloch sphere representation, this Hamiltonian describes a rotation around the axis $\overrightarrow{\Omega}(t) = (\Omega(t)\cos(\phi), -\Omega(t)\sin(\phi), -\delta(t))^T$, with angular velocity $\Omega_{eff}(t) = |\overrightarrow{\Omega}(t)| = \sqrt{\Omega^2(t) + \delta^2(t)}$.

:::{figure} files/bloch_rotation_a_b.png
:align: center
:alt: Representation of the drive Hamiltonian's dynamics as a rotation in the Bloch sphere.
:width: 300

Representation of the drive Hamiltonian's dynamics as a rotation in the Bloch sphere. The coherent excitation is driven between a lower energy level, $|a\rangle$, and a higher energy level,
$|b\rangle$, with Rabi frequency $\Omega(t)$, detuning $\delta(t)$ and phase $\phi$.

> :::

Reviewer: I would hide only this part and call it "Rotations on the Bloch sphere"

:::{important}
With Pulser, you program the driving Hamiltonian by setting $\Omega(t)$, $\delta(t)$ and $\phi(t)$, all the while Pulser ensures that you respect the constraints of your chosen device.
:::

</details>

#### 2.2. Interaction Hamiltonian

<details>

>   <summary>Click to expand</summary>

Reviewer: Again, I wouldn't hide everything, just the details.

$$

- The interaction strength is $\frac{C_3}{R_{ij}^3}$, with $C_3$ a coefficient that depends on the energy levels used to encode $\left|0\right>$ and $\left|1\right>$. 
- The entangling operator between atom $i$ and $j$ is $\hat{\sigma}_i^{+}\hat{\sigma}_j^{-} + \hat{\sigma}_i^{-}\hat{\sigma}_j^{+} = |1\rangle\langle 0|_i |0\rangle\langle 1|_j + |0\rangle\langle 1|_i |1\rangle\langle 0|_j$. 

This interaction hamiltonian is associated with the _XY Hamiltonian_ and is a less common mode of operation, usually accessible only in select neutral-atom devices.

:::{important}
With Pulser, you program the interaction Hamiltonian by setting the distance between atoms, $R_{ij}$. Additionally, the choice of eigenstates used in the computation and the Rydberg level(s) targeted by the device fully determine the interaction strength. Most commonly, the ground and Rydberg states ($\left|g\right>$ and $\left|r\right>$) are used, such that   
$H^\text{int}_{ij} = \frac{C_6}{R_{ij}^6} \hat{n}_i\hat{n}_j$   
When providing the distance between atoms, Pulser ensures that you respect the constraints of your chosen device.

> :::

Reviewer: I think that everything we marked as "important" should not be hidden inside a collapsing section

$$
H(t) = \sum_{i \in targeted\_ atoms} H^D_i(t) + \sum_i \sum_{j<i}H^\text{int}_{ij},
$$

:::{important}
The addressing of a `Channel` defines the number of atoms whose transition will be targeted by a pulse.

The most common addressing for a `Channel` is the `Global` one: all the atoms evolve under the same [driving Hamiltonian](programming.md#21-driving-hamiltonian). 
:::

### 4. Add the Pulses

By adding pulses to a channel, we incrementally construct the [driving Hamiltonian](programming.md#21-driving-hamiltonian):
- Each `Pulse` defines the Rabi frequency $\Omega(t)$, the detuning $\delta(t)$ and the phase $\phi$ of the driving Hamiltonian over a duration $\Delta t$. Similarly, a delay sets all these parameters to zero for a given amount of time.
- The channel dictates the states $\left|a\right>$ and $\left|b\right>$ of the driving Hamiltonian.

By applying a series of pulses and delays, one defines the entire driving Hamiltonian of each atom over time.

## Conclusion

We have successfully defined the [Hamiltonian](programming.md#2-hamiltonian-evolves-the-state) $H$ describing the evolution of the system over time, by:

> - picking a `Device`, that defined the value of the $C_6$ or $C_3$ coefficients.
> - creating a `Register` of atoms, that defined the number of atoms used and the distance between the atoms $R_{ij}$.
> - Selecting the `Channels` of the `Device` to use, that defined the energy levels of the atoms to use: that step completely defined the [interaction Hamiltonian](programming.md#22-interaction-hamiltonian). The addressing property of each `Channel` also dictates the atoms that will be targeted by the `Pulse`.

Reviewer: nit: There should be no comma before "that". I think "which" is more approriate here
```suggestion
- Picking a `Device`, which defines the value of the $C_6$ or $C_3$ coefficients.
- Creating a `Register` of atoms, which defines the number of atoms used and the distance between them, $R_{ij}$.
- Selecting the `Channels` of the `Device` to use, which define the energy levels of the atoms to use - thos step completely defined the [interaction Hamiltonian](programming.md#22-interaction-hamiltonian). The addressing property of each `Channel` also dictates the atoms that will be targeted by the `Pulse`.
```

- Adding `Pulse` and delays to the `Channel`s defines the [driving Hamiltonian](programming.md#21-driving-hamiltonian) of each atom along time.    

You can now simulate your first Hamiltonian by programming your first `Sequence` ! [In this tutorial](tutorials/creating.nblink), you will simulate the evolution of the state of an atom initialized in $\left|g\right>$ under a Hamiltonian $H(t)=\frac{\Omega(t)}{2} |g\rangle \langle r|+\frac{\Omega(t)}{2} |r\rangle\langle g|$, with $\Omega$ chosen such that the final state of the atom is the excited state $\left|r\right>$.

Many concepts have been introduced here and you might want further explanations.

> - The `Device` object contains all the constraints and physical quantities that are defined in a QPU. [This section in the fundamentals](apidoc/core.rst) details these and provides examples of `Devices`. The `VirtualDevices` were also mentioned in this document ([here](programming.md#1-pick-a-device)), this is a most advanced feature described [here](tutorials/virtual_devices.nblink).

Reviewer: ```suggestion
- The `Device` object contains all the constraints and physical quantities that are defined in a QPU. [This section in the fundamentals](apidoc/core.rst) details these and provides examples of `Devices`. The `VirtualDevices` were also mentioned in this document ([here](programming.md#1-pick-a-device)), which is a more advanced feature described [here](tutorials/virtual_devices.nblink).
```

> - There are multiple ways of defining a `Register`, this is further detailed [in this section](tutorials/reg_layouts.nblink).

Reviewer: ```suggestion
- There are multiple ways of defining a `Register`, as is further detailed [in this section](tutorials/reg_layouts.nblink).
```

- The energy levels associated with each `Channel` and the interaction Hamiltonian they implement are summed up in [the conventions page](conventions.md). The channels contain lots of constraints and physical informations, they are detailed in [the same section as the `Device`](apidoc/core.rst).

> - The quantities in a `Pulse` are defined using `Waveform`s, read more about this [on this page](tutorials/composite_wfs.nblink).

Reviewer: ```suggestion
- The quantities in a `Pulse` are defined using `Waveform`s, you read more about these [on this page](tutorials/composite_wfs.nblink).
```
